growcrops.lua

```diff
@@ -0,0 +1,169 @@
+        end
+    end
+
+    return false
+end
+
+-- create a list of available crops to grow, from seeds
+function collectInventory()
+    for _, seed in ipairs(df.global.world.items.other.SEEDS) do
+        -- skip if seed is not in a specific building (plot)
+        if not containsRefType(seed.general_refs, df.general_ref_building_holderst) then
+            goto skipseed
+        end
+
+        if seed.grow_counter < raws_crop[seed.mat_index].growdur then
+            inventory[seed.mat_index] = (inventory[seed.mat_index] or 0) + 1
+        end
+
+        :: skipseed ::
+    end
+end
+
+-- Display a list of planted crops
+function listPlantedCrops()
+    local crops = {}
+    -- create a copy of crops from inventory to use for sorting
+    for mat, invCount in pairs(inventory) do
+        table.insert(crops, { name = raws_crop[mat].name, count = invCount })
+    end
+
+    -- sort the table in descending order by seed count
+    table.sort(crops, function(a, b) return a.count > b.count end)
+
+    for _, crop in ipairs(crops) do
+        print(string.format("%s %d", crop.name, crop.count))
+    end
+end
+
+-- Tries to find a crop in raws_crop that matches the given name
+-- The returned value contains the mat_index and name of the crop
+function findRawsCrop(material)
+    if not material then
+        return nil
+    end
+
+    local possible_matches = {}
+    local lower_mat = string.lower(material)
+    for mat_index, crop in pairs(raws_crop) do
+        local lower_crop_name = string.lower(crop.name)
+        if lower_mat == lower_crop_name then
+            return {mat_index = mat_index, name = crop.name}
+        end
+
+        if string.find(lower_crop_name, lower_mat) then
+            table.insert(possible_matches, {mat_index = mat_index, name = crop.name})
+        end
+    end
+
+    if #possible_matches == 1 then
+        return possible_matches[1]
+    end
+
+    return nil
+end
+
+-- grow specific crop
+function growCrop(material)
+    -- find the matching crop
+    local wanted_mat = findRawsCrop(material)
+    if not wanted_mat then
+        -- no crop with that name
+        qerror("invalid plant material " .. material)
+    end
+
+    -- grow each seed for specified crop
+    local count = 0
+    for _, seed in pairs(df.global.world.items.other.SEEDS) do
+        -- Skip seeds that don't match the material
+        if seed.mat_index ~= wanted_mat.mat_index then
+            goto growskip
+        end
+
+        -- skip if seed is not in a specific building (plot)
+        if not containsRefType(seed.general_refs, df.general_ref_building_holderst) then
+            goto growskip
+        end
+
+        if seed.grow_counter < raws_crop[seed.mat_index].growdur then
+            seed.grow_counter = raws_crop[seed.mat_index].growdur
+            count = count + 1
+        end
+
+        :: growskip ::
+    end
+
+    print(string.format("Grew %d %s", count, wanted_mat.name))
+end
+
+local args = {...}
+local material = args[1]
```

Comment:
You might be better off making this a local instead of caching it across invocations. The world can get unloaded and a new world loaded in its place and then the values in this global will be stale.

Reply:
I think you meant for this comment to be on line 147 `raws_crop = raws_crop or {}`

I'll make it local. The original script seemed to be using it as a global cache between invocations but I'm not sure _why_ they were doing that

Reply:
they *may* just have not considered the world reload case. the simplicity of this script does not warrant caching, though.